jsettlers.graphics/src/main/java/jsettlers/graphics/map/draw/settlerimages/SettlerImageMap.java

```diff
@@ -295,51 +218,31 @@ public Image getImageForSettler(IMovable movable, float progress) {
 				progress += .5f;
 			}
 		}
-		return getImageForSettler(movable.getMovableType(),
-				movable.getAction(), movable.getMaterial(),
-				movable.getDirection(), progress);
+		return getImageForSettler(SettlerImageFlavor.createFromMovable(movable), progress);
 	}
 
 	/**
 	 * Gets an image for a given settler.
 	 *
-	 * @param movableType
-	 * 		The type of the settler.
-	 * @param action
-	 * 		The action the settler is doing.
-	 * @param material
-	 * 		The material that is assigned to the settler.
-	 * @param direction
-	 * 		Its direction.
+	 *
+	 * @param settlerImageFlavor
 	 * @param progress
 	 * 		The progress.
 	 * @return The image.
 	 */
-	public Image getImageForSettler(EMovableType movableType, EMovableAction action, EMaterialType material, EDirection direction, float progress) {
-		SettlerImageMapItem item = getMapItem(movableType, action, material, direction);
-
-		int duration = item.getDuration();
-		int imageIndex;
-		if (duration >= 0) {
-			imageIndex = item.getStart() + Math.min((int) (progress * duration), duration - 1);
-		} else {
-			imageIndex = item.getStart() + Math.max((int) (progress * duration), duration + 1);
-		}
-		return this.imageProvider.getSettlerSequence(item.getFile(), item.getSequenceIndex()).getImageSafe(imageIndex);
+	public Image getImageForSettler(SettlerImageFlavor settlerImageFlavor, float progress) {
+		SettlerImageMapItem item = getMapItem(settlerImageFlavor);
+		return imageProvider.getImageSafe(item, progress);
 	}
 
 	/**
 	 * Gets a map item.
 	 *
-	 * @param movableType
-	 * @param action
-	 * @param material
-	 * @param direction
-	 * @param progress
-	 * @return The item of the map at the given position. Is not null.
+	 *
+	 * @param settlerImageFlavor@return The item of the map at the given position. Is not null.
 	 */
-	private SettlerImageMapItem getMapItem(EMovableType movableType, EMovableAction action, EMaterialType material, EDirection direction) {
-		SettlerImageMapItem item = this.map[movableType.ordinal()][action.ordinal()][material.ordinal][direction.ordinal];
+	private SettlerImageMapItem getMapItem(SettlerImageFlavor settlerImageFlavor) {
+		SettlerImageMapItem item = map.get(settlerImageFlavor);
```

Comment:
Did you time the old and the new retrieval? I guess the performance of accessing a hash map vs. accessing a 4 dim array might be worse. How much is it for a larger number of calls (e.g. 1000 or so)?

Reply:
Actually, accessing a hash map should be faster than a 4 dimensional array (fewer memory accesses / cache misses).

Reply:
Access of hash map by key is 0(1) so the same as the existing index lookup.

Memory footprint is far less because the array was as big as all possible combinations. I measured over a million. By using a hash map it went down to a little more than 1000. The expressiveness and the lower memory footprint were the 2 major reason why I did this refactoring.

Reply:
Actually the HashMap implementation of java has no guaranteed access of O(1) but a worst case access of O(n) instead ;-). But we don't care about worst case complexity here, we care about actual performance. Measuring this is hard - you cannot measure the GC delays you get in a few frames so easily. So you need to track speed and GC consumption with both the old and new methods in a reproducible setting (should be enough to test it on PC though, Android is even more complex for this).

Reply:
You said that the HashMap should be faster than the array. I'll look again at the hash function inside the flavor class and take care about object creation inside drawing methods to avoid GC pressure. Your two comments seem to contradict?

The array removal was merely a read- and maintainability change plus a huge memory gain.